src/methods.ts

```diff
@@ -226,6 +227,98 @@ export class Field extends AbstractMethodField {
   }
 }
 
+const opcodeSize: number[] = function() {
+  const table:number[] = [];
+  const layoutType = enums.OpcodeLayoutType;
+
+  table[layoutType.OPCODE_ONLY] = 1;
+  table[layoutType.CONSTANT_POOL_UINT8] = 2;
+  table[layoutType.CONSTANT_POOL] = 3;
+  table[layoutType.CONSTANT_POOL_AND_UINT8_VALUE] = 4;
+  table[layoutType.UINT8_VALUE] = 2;
+  table[layoutType.UINT8_AND_INT8_VALUE] = 3;
+  table[layoutType.INT8_VALUE] = 2;
+  table[layoutType.INT16_VALUE] = 3;
+  table[layoutType.INT32_VALUE] = 5;
+  table[layoutType.ARRAY_TYPE] = 2;
+  table[layoutType.WIDE] = 1;
+
+  return table;
+}();
+
+class TraceInfo {
+  pops: string[] = [];
+  pushes: string[] = [];
+  prefixEmit: string = "";
+  onErrorPushes: string[];
+
+  constructor(public pc: number, public jitInfo: JitInfo) {
+  }
+}
+
+class Trace {
+  private infos: TraceInfo[] = [];
+
+  constructor(public startPC: number, private code: Buffer, private method: Method) {
+  }
+
+  public addOp(pc: number, jitInfo: JitInfo) {
+    this.infos.push(new TraceInfo(pc, jitInfo));
+  }
+
+  public close(): Function {
+    if (this.infos.length > 1) {
+      const symbolicStack: string[] = [];
+      let symbolCount = 0;
+      let emitted = "";
+      for (let i = 0; i < this.infos.length; i++) {
+        const info = this.infos[i];
+        const jitInfo = info.jitInfo;
+
+        const pops = info.pops;
+        const normalizedPops = jitInfo.pops < 0 ? Math.min(-jitInfo.pops, symbolicStack.length) : jitInfo.pops;
+        for (let j = 0; j < normalizedPops; j++) {
+          if (symbolicStack.length > 0) {
+            pops.push(symbolicStack.pop());
+          } else {
+            const symbol = "s" + symbolCount++;
+            info.prefixEmit += `var ${symbol} = f.opStack.pop();`;
+            pops.push(symbol);
+          }
+        }
+
+        info.onErrorPushes = symbolicStack.slice();
+
+        const pushes = info.pushes;
+        for (let j = 0; j < jitInfo.pushes; j++) {
+          const symbol = "s" + symbolCount++;
+          symbolicStack.push(symbol);
+          pushes.push(symbol);
+        }
+
+      }
+
+      if (symbolicStack.length === 1) {
+        emitted += `f.opStack.push(${symbolicStack[0]});`;
+      } else if (symbolicStack.length > 1) {
+        emitted += `f.opStack.pushAll(${symbolicStack.join(',')});`;
+      }
+
+      for (let i = this.infos.length-1; i >= 0; i--) {
+        const info = this.infos[i];
+        const jitInfo = info.jitInfo;
+        emitted = info.prefixEmit + jitInfo.emit(info.pops, info.pushes, ""+i, emitted, this.code, info.pc, info.onErrorPushes, this.method);
+      }
+
+      // console.log(`Emitted trace of ${this.infos.length} ops: ` + emitted);
+      // f = frame, t = thread, u = util
+      return new Function("f", "t", "u", emitted);
```

Comment:
Indeed, `eval` has better debugging support. I chose `new Function()` for its performance, after a lot of benchmarking. You can see the results and benchmarking code [here](http://git.javadeploy.net/hrjet9/doppio-jit/issues/1). (It is not super descriptive; it was mainly intended for my own reference. To give a brief summary: I implemented a mini interpreter and various versions of a JITed function to compare their speed across Chromium, Firefox and node).

For debugging, I agree, `new Function()` has poor support. Though, Chromium does assign it a name automatically, with clickable links, if there is an exception. And Firefox has a non-standard feature called [displayName](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/displayName) that helps.